core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java

```diff
@@ -257,10 +257,11 @@ private CompletionStage<CacheStatusResponse> sendJoinRequest(String cacheName, C
             log.debugf("Join request received CacheNotFoundResponse for cache %s, retrying", cacheName);
          } else {
             log.debugf(t, "Join request failed for cache %s", cacheName);
-            if (t instanceof TimeoutException) {
-               throw (TimeoutException) t;
+            if (t.getCause() instanceof CacheJoinException) {
+               throw (CacheJoinException) t.getCause();
+            } else {
+               throw new CacheException(t);
```

Comment:
Check my comment at https://github.com/infinispan/infinispan/pull/12163/files#r1537704092

Reply:
Going to update it here.